packages/admin-panel-server/src/viz-builder/dashboardVisualisation/DashboardVisualisationExtractor.ts

```diff
@@ -5,43 +5,53 @@
 
 import { isNil, omitBy } from 'lodash';
 
-import { snakeKeys } from '@tupaia/utils';
-import {
-  DashboardItem,
-  DashboardVisualisationObject,
-  DashboardVisualisationResource,
-  PreviewMode,
-  Report,
-  CamelKeysToSnake,
-  VisualisationValidator,
-} from '../types';
+import { snakeKeys, yup } from '@tupaia/utils';
+import { PreviewMode, CamelKeysToSnake } from '../types';
 
-export class DashboardVisualisationExtractor {
-  private readonly visualisation: DashboardVisualisationObject;
+import { baseVisualisationValidator } from './validators';
 
-  private readonly dashboardItemValidator: VisualisationValidator;
+// expands object types recursively
+// TODO: Move this type to a generic @tupaia/utils-ts package
+type ExpandType<T> = T extends Record<string, unknown>
+  ? T extends infer O
+    ? {
+        [K in keyof O]: ExpandType<O[K]>;
+      }
+    : never
+  : T;
 
-  private readonly reportValidator: VisualisationValidator;
+export class DashboardVisualisationExtractor<
+  DashboardItemValidator extends yup.AnyObjectSchema,
+  ReportValidator extends yup.AnyObjectSchema
+> {
+  private readonly visualisation: ExpandType<yup.InferType<typeof baseVisualisationValidator>>;
+
+  private readonly dashboardItemValidator: DashboardItemValidator;
+
+  private readonly reportValidator: ReportValidator;
 
   constructor(
-    visualisation: DashboardVisualisationObject,
-    dashboardItemValidator: VisualisationValidator,
-    reportValidator: VisualisationValidator,
+    visualisation: Record<string, unknown>,
+    dashboardItemValidator: DashboardItemValidator,
+    reportValidator: ReportValidator,
   ) {
-    this.visualisation = visualisation;
+    this.visualisation = baseVisualisationValidator.validateSync(visualisation);
     this.dashboardItemValidator = dashboardItemValidator;
     this.reportValidator = reportValidator;
   }
 
-  public extractDashboardVisualisationResource = (): DashboardVisualisationResource => {
+  public getDashboardVisualisationResource = () => {
     // Resources (like the ones passed to meditrak-server for upsert) use snake_case keys
-    const dashboardItem = snakeKeys(this.getDashboardItem()) as CamelKeysToSnake<DashboardItem>;
-    const report = snakeKeys(this.getReport()) as CamelKeysToSnake<Report>;
+    const dashboardItem = this.getDashboardItem();
+    const report = this.getReport();
 
-    return { dashboardItem, report };
+    return { dashboardItem: snakeKeys(dashboardItem), report: snakeKeys(report) } as {
+      dashboardItem: ExpandType<CamelKeysToSnake<typeof dashboardItem>>;
```

Comment:
I believe we don't need to expand the `CamelKeysToSnake` type here, since it is only applied to the top level of `dashboardItem` fields

Reply:
Is it not needed to convert `reportCode` to `report_code` in the type? I think I might be confused...

Reply:
I believe that `CamelKeysToSnake` is enough to do that, but I may be wrong?

Reply:
It's nothing major, but it changes the return type from showing up in vs code like this:
![image](https://user-images.githubusercontent.com/59544282/131941905-02b20a70-3288-49f3-8ad8-6718f0ec90dd.png)

to this:
![image](https://user-images.githubusercontent.com/59544282/131941952-a6975bda-3fc8-43f6-9c50-dd7f6cc08156.png)

Reply:
Oh nice, let's keep it then!